pilot/pkg/config/kube/gateway/inferencepool_collection.go

```diff
@@ -0,0 +1,319 @@
+	hashSize                      = 8
+	InferencePoolRefLabel         = "inference.x-k8s.io/inference-pool-name"
+	InferencePoolExtensionRefSvc  = "inference.x-k8s.io/extension-service"
+	InferencePoolExtensionRefPort = "inference.x-k8s.io/extension-port"
+)
+
+// // ManagedLabel is the label used to identify resources managed by this controller
+// const ManagedLabel = "inference.x-k8s.io/managed-by"
+
+// ControllerName is the name of this controller for labeling resources it manages
+const ControllerName = "inference-controller"
+
+var supportedControllers = getSupportedControllers()
+
+func getSupportedControllers() sets.Set[gatewayv1.GatewayController] {
+	ret := sets.New[gatewayv1.GatewayController]()
+	for _, controller := range builtinClasses {
+		ret.Insert(controller)
+	}
+	return ret
+}
+
+type shadowServiceInfo struct {
+	key        types.NamespacedName
+	selector   map[string]string
+	poolName   string
+	poolUID    types.UID
+	targetPort int32
+}
+
+type extRefInfo struct {
+	name string
+	port int32
+}
+
+type InferencePool struct {
+	shadowService shadowServiceInfo
+	extRef        extRefInfo
+}
+
+func (i InferencePool) ResourceName() string {
+	return i.shadowService.key.Namespace + "/" + i.shadowService.poolName
+}
+
+func InferencePoolCollection(
+	pools krt.Collection[*inferencev1alpha2.InferencePool],
+	services krt.Collection[*corev1.Service],
+	httpRoutes krt.Collection[*gateway.HTTPRoute],
+	gateways krt.Collection[*gateway.Gateway],
+	routesByNamespace krt.Index[string, *gateway.HTTPRoute],
+	c *Controller,
+	opts krt.OptionsBuilder,
+) (krt.StatusCollection[*inferencev1alpha2.InferencePool, inferencev1alpha2.InferencePoolStatus], krt.Collection[InferencePool]) {
+	return krt.NewStatusCollection(pools,
+		func(
+			ctx krt.HandlerContext,
+			pool *inferencev1alpha2.InferencePool,
+		) (*inferencev1alpha2.InferencePoolStatus, *InferencePool) {
+			// First, let's build the shadow service
+			extRef := extRefInfo{
+				name: string(pool.Spec.ExtensionRef.Name),
+			}
+			if pool.Spec.ExtensionRef.PortNumber != nil {
+				extRef.port = int32(*pool.Spec.ExtensionRef.PortNumber)
+			} else {
+				extRef.port = 9002 // Default port for the inference extension
+			}
+
+			svcName, err := InferencePoolServiceName(pool.Name)
+			if err != nil {
+				log.Errorf("failed to generate service name for InferencePool %s: %v", pool.Name, err)
+				return nil, nil
+			}
+
+			shadowSvcInfo := shadowServiceInfo{
+				key: types.NamespacedName{
+					Name:      svcName,
+					Namespace: pool.GetNamespace(),
+				},
+				selector:   make(map[string]string, len(pool.Spec.Selector)),
+				poolName:   pool.GetName(),
+				targetPort: pool.Spec.TargetPortNumber,
+				poolUID:    pool.GetUID(),
+			}
+
+			for k, v := range pool.Spec.Selector {
+				shadowSvcInfo.selector[string(k)] = string(v)
+			}
+
+			// Make sure we reconcile the service if it is manually changed
+			_ = krt.FetchOne(ctx, services, krt.FilterKey(shadowSvcInfo.key.String()))
```

Comment:
Not sure I get this. It won't impact the outcome so it won't call the handler?

Likely this was left over from before

Reply:
Yeah but running the handler again does nothing. It will produce the same output, and when then the event handler will never be called. krt de-dupes outputs that are equal

Reply:
Actually decided against reconciling since users may want to label the inferencepool service

Reply:
I think we should reconcile, if they changed the labels they can mess up their pool. We should en up doing patch and not update thought - which is a separate effort